File: ProjectEuler/P3.py
#!/usr/bin/env python
"""
The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 317584931803?
"""

from math import sqrt
from itertools import takewhile

def primes(n):
    # From python cookbook
    if n==2: return [2]
    elif n<2: return []
    s=range(3,n+1,2)
    mroot = n ** 0.5
    half=(n+1)/2-1
    i=0
    m=3
    while m <= mroot:
        if s[i]:
            j=(m*m-3)/2
            s[j]=0
            while j<half:
                s[j]=0
                j+=m
        i=i+1
        m=2*i+3
    return [2]+[x for x in s if x]

if __name__ == '__main__': 
    n = 317584931803
    imax = int(sqrt(n))

    for i in primes(imax):
        if n % i == 0: print i

File: ProjectEuler/P3.cc
//The prime factors of 13195 are 5, 7, 13 and 29. What is the largest
//prime factor of the number 317584931803?

#include <cmath>
#include <iostream>
#include <vector>
using namespace std;

vector<int> primes(unsigned long long nmax){
  // Construct a vector of primes using a sieve of Eratosthenes
  int i=0,j=0;
  vector<bool> isprime;
  vector<int> ps;
  int m, mroot;
  for (i=0; i<=nmax; i++) 
    isprime.push_back(true);
  isprime[0] = false;
  isprime[1] = false;
  for (i=2; i*i <= nmax; i++){
    if (isprime[i]){
      for (j=2; i*j <= nmax; j++)
	isprime[i*j] = false;
    }
  }
  for (i=1; i<nmax; i++){
    if (isprime[i]) {
      ps.push_back(i);
    }
  }
  return ps;
}

main(){
  unsigned long long num = 317584931803ULL;
  long imax = sqrt(num);
  vector<int> ps;
  ps = primes(imax);
  // print out the primes, if desired:
  //for (vector<int>::iterator i = ps.begin(); i < ps.end(); i++)
  //  cout << *i << " ";
  //cout << endl;
  for (vector<int>::iterator i = ps.begin(); i < ps.end(); i++){
    if (num % *i == 0) cout << *i << endl;
  }
}
